src/interval.c

```diff
@@ -256,31 +256,65 @@ ts_interval_from_now_func_get_datum(int64 interval, Oid time_dim_type, Oid now_f
 	switch (time_dim_type)
 	{
 		case INT2OID:
-			return Int64GetDatum(DatumGetInt16(now) - interval);
+			return Int16GetDatum(DatumGetInt16(now) - interval);
 		case INT4OID:
-			return Int64GetDatum(DatumGetInt32(now) - interval);
+			return Int32GetDatum(DatumGetInt32(now) - interval);
 		case INT8OID:
 			return Int64GetDatum(DatumGetInt64(now) - interval);
 		default:
 			pg_unreachable();
 	}
 }
 
+static bool
+noarg_integer_now_func_filter(Form_pg_proc form, void *arg)
+{
+	Oid *rettype = arg;
+
+	return form->pronargs == 0 && form->prorettype == *rettype;
+}
+
+/* maybe this can be exported later if other parts of the code need
+ * to access the integer_now_func
+ */
+static Oid
+get_integer_now_func(Dimension *open_dim)
+{
+	Oid rettype;
+	Oid now_func;
+
+	rettype = ts_dimension_get_partition_type(open_dim);
+
+	Assert(IS_INTEGER_TYPE(rettype));
+
+	now_func = lookup_proc_filtered(NameStr(open_dim->fd.integer_now_func_schema),
+									NameStr(open_dim->fd.integer_now_func),
+									NULL,
+									noarg_integer_now_func_filter,
+									&rettype);
+	return now_func;
+}
+
 /*
  * Convert the difference of interval and current timestamp to internal representation
```

Comment:
please update the comment to reflect that the return value is a datum of the dimension type